### PR TITLE
[Store] add codec inference and recursive structure expansion

### PR DESCRIPTION
**Pull request by zxpdemonio**
### Background & Motivation

  In RLHF (Reinforcement Learning from Human Feedback) training pipelines, the actor and learner stages exchange rollout data through mooncake store. A single rollout batch (typically represented as a `DataProto` object) contains  heterogeneous fields: dense tensors for token IDs and attention masks, ragged tensors for variable-length responses, byte arrays for serialized reward signals, strings for prompts, JSON metadata, and nested dicts/lists that group these fields hierarchically.

  Today, the caller must manually serialize each field, choose an appropriate wire format, and manage the Mooncake put/get lifecycle per field. This is error-prone, verbose, and prevents Mooncake from applying type-specific optimizations (e.g., zero-copy tensor transfer via registered buffers vs. pickle fallback for opaque Python objects).
  
  PR #2050 introduced a complete DataProto transfer framework that automates this: it inspects the runtime structure of a rollout batch, recursively decomposes nested dicts/lists into flat leaf columns, selects the best codec for each leaf, and encodes/decodes them through Mooncake's structured object store. That PR was large (~3000 lines) and is being split into reviewable pieces by functional layer.

  ### What this PR does (PR 3 of the #2050 series)

  This PR adds the **first two layers** of the encoding pipeline — the parts that run before any actual serialization or I/O:

  1. **Recursive structure expansion** (`infer_structure`): Given a batch of N rows where each row is an arbitrary Python value (dict, list, tensor, string, etc.), recursively decomposes dict and short-list rows into child columns. For example, N rows of `{"tokens": tensor, "meta": {"step": int, "lr": float}}` become three flat leaf columns: `tokens`, `meta.step`, and `meta.lr`. Interior nodes (dict expansions, list expansions) are recorded for later reconstruction.

  2. **Codec inference** (`_choose_leaf_codec`): For each leaf column (a list of N homogeneous-typed values), selects the optimal storage codec by trying type predicates in priority order: `ragged_tensor` → `media_list_ragged` → `typed_ragged` → `ndarray` → `bytes_ragged` → `media_bytes` → `utf8_ragged` → `json_ragged` → `pickle_ragged_fallback`.
  The decision includes the codec name and any type metadata (e.g., tensor dtype, numpy result dtype) needed by the encoder in the next PR.

  ### What this PR does NOT do

  - No actual encoding/decoding (PR 4)
  - No integration into `put_structured_object` / `materialize` (PR 5)
  - No store I/O, no buffer registration, no network calls

  This is a pure-function layer with no side effects, making it independently testable and reviewable.

  ### Split plan from #2050
  
  - **PR 1** — Buffer pool (local buffer backing) — *Merged*
  - **PR 2** — Structured object interfaces (tensor APIs) — *Merged*
  - **PR 3 (this)** — **Codec inference + recursive structure expansion** — *This PR*
  - **PR 4** — Leaf encoders/decoders (ragged tensor, typed ragged, bytes, media, text, json, pickle) — *Planned*
  - **PR 5** — Integration into put/materialize paths + end-to-end DataProto transfer — *Planned*

  ### Design decisions

  - **Expansion before codec inference**: `infer_structure` tries dict/list expansion *before* calling
  `_choose_leaf_codec`. Otherwise, `[{"x": 1}, {"x": 2}]` would be accepted as `json_ragged` instead of being expanded into a scalar column `x` with `ndarray` codec — losing the opportunity for zero-copy numeric transfer.
  - **`_try_expand_dict` / `_try_expand_list` return computed data**: The expansion check and the actual expansion need the same derived values (sorted keys, max length, per-row lengths). These functions return the computed data on success (`list[str]` or `(int, list[int])`) and `None` on rejection, avoiding a second traversal in `infer_structure`.
  - **`_check_all` helper**: Five of the eight codec predicates follow an identical "filter nulls → type-check all → accept/reject" pattern. `_check_all` factors this out; the remaining three predicates (`_can_tensor`, `_can_numeric_sequence`, `_can_numeric_scalar`) have extra logic (dtype uniformity check, `np.result_type` promotion) and stay as standalone functions.
  - **Recursion depth limit**: Capped at 32 levels with a descriptive `ValueError`. Prevents stack overflow on pathological inputs.
  - **Internal symbols**: All new names except `infer_structure` are underscore-prefixed — `_CodecDecision`,
  `_InferredLeaf`, `_InferredNode`, `_choose_leaf_codec`, etc.

  ## Module
  
  - [x] Python Wheel (`mooncake-wheel`)

  ## Type of Change

  - [x] New feature

  ## How Has This Been Tested?

  **Test commands:**
  ```bash
  PYTHONPATH=mooncake-wheel python -m pytest tests/test_put_get_tensor.py::TestCodecInference -v
  ```
  Test results (18 tests):

  Codec selection:
  - tensor, mixed-dtype rejection, numeric sequence, bytes, text, json, scalar, fallback, with-nulls, empty input, all-None

  Structure expansion:
  - flat leaf, dict expansion, dict with None rows, nested dict, list expansion with full path/lengths verification, depth limit
  - [x] Unit tests pass
  - [x] No store or network dependency — runs in pure Python + numpy + torch-cpu

  Checklist

  - [x] I have performed a self-review of my own code
  - [x] I have added tests to prove my changes are effective                                           
  - [ ] For changes >500 LOC: I have filed an RFC issue

  AI Assistance Disclosure

  - [x] AI tools were used (specify below)
  
  Claude Opus 4.6 assisted with code generation, review (3 parallel agents covering simplicity/safety/performance), and test writing. All changes reviewed and validated by the human submitter.

### CODE REVIEW

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 1]
```diff
@@ -1303,3 +1303,228 @@ def _cleanup_keys(store: BundleStore, keys: Sequence[str], strict: bool) -> None
...
+
+    *leaves* and *nodes* are output accumulators populated during recursion.
+    """
+    if _depth > _INFER_MAX_DEPTH:
+        raise ValueError(f"infer_structure exceeded max depth {_INFER_MAX_DEPTH} at {path!r}")
+    dict_keys = _try_expand_dict(values)
+    if dict_keys is not None:
+        nodes.append(_InferredNode(path, "dict", dict_keys))
+        for key in dict_keys:
+            child = [v.get(key) if isinstance(v, dict) else None for v in values]
```
Comment (yokinoshitayoki) [inline mooncake-wheel/mooncake/structured_object_store.py]: I'm afraid this loses information needed for later reconstruction. `v.get(key)` collapses three different cases into the same `None` child value: the key is present with value `None`, the key is missing, and the parent row itself is `None`. Since the PR says interior nodes are recorded for later reconstruction, the inferred node metadata needs to preserve parent presence and per-key presence, or nullable/sparse dicts should not be expanded.

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 2]
```diff
@@ -1303,3 +1303,228 @@ def _cleanup_keys(store: BundleStore, keys: Sequence[str], strict: bool) -> None
...
+            infer_structure(f"{path}.{key}", child, leaves, nodes, _depth=_depth + 1)
```
Comment (yokinoshitayoki) [inline mooncake-wheel/mooncake/structured_object_store.py]: Path construction is ambiguous here. Because dict keys are arbitrary strings, inputs like `{"a.b": 1, "a": {"b": 2}}` produce duplicate leaf paths (`r.a.b`). Keys containing `[0]` can also collide with list-index paths. If these paths are later used as column names or metadata keys, reconstruction can overwrite or mix fields. Consider representing paths as structured tokens internally, or escaping delimiters and rejecting duplicate leaf paths.

File: mooncake-wheel/mooncake/structured_object_store.py (hunk 1 above)
Comment (zxpdemonio) [reply mooncake-wheel/mooncake/structured_object_store.py]: Fixed in the latest push.

The three cases are now distinguishable:
 - Key present, value is None: child column stores None, row_mask[i] = True
 - Key absent (sparse dict): child column stores a MISSING sentinel object
  - Parent row is None: child column stores None, row_mask[i] = False

_InferredNode now carries a row_mask: list[bool] that records which rows had a real parent container. _non_null filters both None and MISSING so codec predicates are unaffected.
The decoder (in a follow-up PR) can use row_mask + MISSING to faithfully reconstruct sparse/nullable dicts without losing information.

File: mooncake-wheel/mooncake/structured_object_store.py (hunk 2 above)
Comment (zxpdemonio) [reply mooncake-wheel/mooncake/structured_object_store.py]: Agreed — this was a real collision risk. Fixed by escaping dict keys in paths: \ → \\, . → \., [ → \[.

A key like "a.b" now produces path r.a\.b, which is unambiguous from the nested dict path r.a.b. The unescape function will be added in the follow-up PR that handles reconstruction. List index paths ([0], [1], ...) are always numeric and never collide with escaped keys.

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 3]
```diff
@@ -1303,3 +1303,248 @@ def _cleanup_keys(store: BundleStore, keys: Sequence[str], strict: bool) -> None
...
+        return _CodecDecision(False, "ndarray", f"non-numeric dtype: {dtype}", "numeric scalar")
+    return _CodecDecision(True, "ndarray", "all rows are numeric scalar", "numeric scalar", {"dtype": str(dtype)})
+
+
+def _can_json(values: list[Any]) -> _CodecDecision:
+    nn = _non_null(values)
+    if not nn:
+        return _CodecDecision(False, "json_ragged", "all rows are null", "json")
+    total_bytes = 0
+    for v in nn[:_INFER_MAX_SAMPLE_ROWS]:
```
Comment (yokinoshitayoki) [inline mooncake-wheel/mooncake/structured_object_store.py]: This still only validates JSON serializability for the first 128 non-null rows. A column like `[{"ok": 1}] * 128 + [object()]` would be inferred as `json_ragged`, but the actual encoder would fail when it reaches the later non-serializable row. Codec inference should only select the JSON codec if every non-null row is JSON-serializable; the byte-size limit can be sampled, but serializability should be checked across the full column.

Comment (zxpdemonio) [reply mooncake-wheel/mooncake/structured_object_store.py]: Fixed. _can_json now validates JSON serializability across all non-null rows, not just the first 128. The byte-size estimate is still sampled (first 128 rows) since it only guards against oversized payloads and doesn't need to be exact. A non-serializable value at any row index will now correctly reject the column.

**Comment by yokinoshitayoki** [comment]
Should we add a regression test for a non-JSON-serializable value after the sampled prefix?

**Comment by stmatengss** [comment]
@qoder Review it by line

**Comment by zxpdemonio** [comment]
> Should we add a regression test for a non-JSON-serializable value after the sampled prefix?

Added. test_json_rejects_late_non_serializable creates a column with 200 valid JSON rows followed by one object(), and asserts the codec is rejected. This guards against the original sampling-only bug.

File: mooncake-wheel/mooncake/structured_object_store.py [hunk 4]
```diff
@@ -1303,3 +1303,250 @@ def _cleanup_keys(store: BundleStore, keys: Sequence[str], strict: bool) -> None
...
+    lengths = [len(v) if isinstance(v, (list, tuple)) else 0 for v in values]
+    return max_len, lengths
+
+
+def _escape_key(key: str) -> str:
+    return key.replace("\\", "\\\\").replace(".", "\\.").replace("[", "\\[")
+
+
+def infer_structure(
+    path: str,
```
[the remark this answers is not in the document]
Comment (zxpdemonio) [reply mooncake-wheel/mooncake/structured_object_store.py:1515]: Thanks for the thorough review. Agreed on all points — no changes needed.
